lib/realtime/ghost.ex

```diff
@@ -104,7 +104,8 @@ defmodule Realtime.Ghost do
                   nil
                 end,
               scheduled_timepoint_status: timepoint_status,
-              route_status: route_status
+              route_status: route_status,
+              block_waivers: BlockWaiver.block_waivers_for_block(block)
```

Comment:
Could you please say more about what you mean by this? I'm not sure I'm following at all.

Reply:
To make a `%Vehicle` from the data from swiftly and busloc VehiclePositions, the VPs get passed as arguments directly from the `consumer/VPs.ex` to `vehicle.ex`.

The Blockwaivers get fetched from a very parallel process, so I wonder why they have to take this roundabout way to get into a Vehicle, where they split paths from the VPs, go through a separate consumer and a gen server, and then get queried back into the vehicle deep in the process.

This adds complexity, because you need to make the new processes, set up all the message passing, and mock any global state you want when testing. We should look for a way to do it with vanilla function arguments instead if possible.

Right now `consumer/VPs.ex` sees all the merge groups, and then throws out anything that's not a VP. Perhaps it could keep the block waivers, and then pass them into `Vehicle.from_vp()`, so the block waivers always stay in the same process as the vps, without any new processes or message passing required.

Reply:
This feels like a lot of intermingling of responsibility and data. I think having very "pure" and simple consumers (`VehiclePositions` and `StopTimeUpdates`) feels very understandable. If anything, your question is making me think that `VehiclePositions` current has too much responsibility it shouldn't (but I don't want to refactor that in this PR).

Reply:
Keeping the code separate for block waivers and vehicle positions is good. According to the [docs](https://hexdocs.pm/gen_stage/GenStage.html#module-usage-guidelines), the best practice is to use modules to keep the logic separate, and that the processes should be combined when the data can flow together.

> Use processes/stages to model runtime properties, such as concurrency and data-transfer, and not for code organization or domain design purposes. For the latter, you should use modules and functions.

I think you're right that `VehiclePositions` has too much responsibility, and if it consumes the block waiver trip updates too, then the name won't be right any more. I'd be okay with a small refactor to rename things and better separate the domain logic into modules. I think the amount and complexity of code in that small refactor would be less than the code we save by not having to juggle more processes, so it wouldn't be inflating this PR.